File: Sample/Java/src/ScorpioProto/Commons/ScorpioReader.java
package ScorpioProto.Commons;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Calendar;
public class ScorpioReader implements IScorpioReader {
	ByteBuffer reader;
	public ScorpioReader(byte[] buffer) {
		reader = ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN);
	}
	public boolean ReadBool() {
		return ReadInt8() == (byte) 1;
	}
	public byte ReadInt8() {
		return reader.get();
	}
	public short ReadInt16() {
		return reader.getShort();
	}
	public int ReadInt32() {
		return reader.getInt();
	}
	public long ReadInt64() {
		return reader.getLong();
	}
	public float ReadFloat() {
		return reader.getFloat();
	}
	public double ReadDouble() {
		return reader.getDouble();
	}
	public String ReadString() {
		try {
			int length = ReadInt32();
			byte[] bytes = new byte[length];
			reader.get(bytes);
			return new String(bytes, "utf-8");
		} catch (Exception e) {
			
		}
		return "";
	}
	public Calendar ReadDateTime() {
		Calendar calendar = Calendar.getInstance();
		calendar.setTimeInMillis(ReadInt64());
		return calendar;
	}
	public byte[] ReadBytes() {
		int length = reader.getInt();
		byte[] bytes = new byte[length];
		reader.get(bytes);
		return bytes;
	}
}

File: Sample/ts/src/ScorpioProto/Commons/ScorpioReader.ts
import { IScorpioReader } from './IScorpioReader';
import { Buffer } from 'buffer'; 
import Long from 'long'
export class ScorpioReader implements IScorpioReader {
    offset:number = 0;
    buffer:Buffer;
    constructor(buf:any) {
        this.offset = 0
        this.buffer = Buffer.from(buf);
    }
    ReadBool():boolean {
        return this.ReadInt8() == 1;
    }
    ReadInt8():number {
        let ret = this.buffer.readInt8(this.offset);
        this.offset += 1
        return ret
    }
    ReadInt16():number {
        let ret = this.buffer.readInt16LE(this.offset);
        this.offset += 2
        return ret
    }
    ReadInt32():number {
        let ret = this.buffer.readInt32LE(this.offset);
        this.offset += 4
        return ret
    }
    ReadInt64():any {
        let low = this.ReadInt32()
        let high = this.ReadInt32()
        return Long.fromBits(low, high, false)
    }
    ReadFloat():number {
        let ret = this.buffer.readFloatLE(this.offset);
        this.offset += 4
        return ret
    }
    ReadDouble():number {
        let ret = this.buffer.readDoubleLE(this.offset);
        this.offset += 8
        return ret
    }
    ReadString():string {
        let length = this.ReadInt32();
        let start = this.offset
        let end = start + length
        let ret = this.buffer.toString("utf8", start, end)
        this.offset += length
        return ret
    }
    ReadDateTime():Date {
        let time = this.ReadInt64()
        return new Date(time.toNumber())
    }
}